src/java.base/share/classes/sun/security/ssl/SSLTransport.java

```diff
@@ -105,44 +106,62 @@ static Plaintext decode(TransportContext context,
         ByteBuffer[] srcs, int srcsOffset, int srcsLength,
         ByteBuffer[] dsts, int dstsOffset, int dstsLength) throws IOException {
 
-        Plaintext[] plaintexts;
-        try {
-            plaintexts =
-                    context.inputRecord.decode(srcs, srcsOffset, srcsLength);
-        } catch (UnsupportedOperationException unsoe) {         // SSLv2Hello
-            // Code to deliver SSLv2 error message for SSL/TLS connections.
-            if (!context.sslContext.isDTLS()) {
-                context.outputRecord.encodeV2NoCipher();
-                if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
-                    SSLLogger.finest("may be talking to SSLv2");
-                }
+        Plaintext[] plaintexts = null;
+
+        // Check for unexpected plaintext alert message during TLSv1.3 handshake, @bug 8331682
+        if (srcsLength == 1 && context.handshakeContext != null &&
```

Comment:
The unencrypted message may only be permitted right after the ClientHello; we need a better check for that. `handshakeContext` will be non-null after handling a key_update message, for example.

Reply:
That's a good suggestion, thanks! I'm adding an additional `!context.isNegotiated` check. `isNegotiated` should be always true after the handshake. The `handshakeContext` after handling a `key_update` message would be of `PostHandshakeContext` type, we can also check for that but it seems redundant.